res/res_pjsip_session.c

```diff
@@ -1115,6 +1230,70 @@ static int handle_negotiated_sdp(struct ast_sip_session *session, const pjmedia_
 		SCOPE_EXIT_RTN_VALUE(-1, "Media stream count mismatch\n");
 	}
 
+	if (pjmedia_sdp_neg_was_answer_remote(session->inv_session->neg)) {
+		int index;
+		for (index = 0; index < local->media_count; index++) {
+			RAII_VAR(struct ast_format_cap *, remote_caps, NULL, ao2_cleanup);
+			RAII_VAR(struct ast_stream_topology *, remote_top, NULL, ao2_cleanup);
+			struct ast_stream *remote_ast_stream;
+			struct ast_stream *resolved_stream;
+			struct ast_stream *configured_stream = ast_stream_topology_get_stream(session->pending_media_state->topology, index);
+			int skip_stream_setting = 0;
+
+			if (!remote->media[index]) {
+				ast_stream_set_state(configured_stream, AST_STREAM_STATE_REMOVED);
+				continue;
+			}
+
+			if (!(remote_caps = ast_format_cap_alloc(AST_FORMAT_CAP_FLAG_DEFAULT))) {
+				ast_log(LOG_WARNING, "%s: Couldn't allocate remote capabilities.\n",
+					ast_sip_session_get_name(session));
+				return -1;
+			}
+			set_caps_from_sdp(session, remote->media[index], remote_caps, 1);
+			if (ast_format_cap_empty(remote_caps)) {
+				continue;
+			}
+
+			remote_top = ast_stream_topology_create_from_format_cap(remote_caps);
+			if (!remote_top || ast_stream_topology_get_count(remote_top) < 1) {
+				ast_log(LOG_WARNING, "%s: Couldn't create remote topology\n",
+					ast_sip_session_get_name(session));
+				return -1;
+			}
+
+			remote_ast_stream = ast_stream_topology_get_stream(remote_top, 0);
+			resolved_stream = ast_stream_create_resolved(
+				remote_ast_stream,
+				configured_stream,
+				&session->endpoint->media.codec_prefs_incoming_answer,
+				NULL);
+			if (!resolved_stream || ast_format_cap_empty(ast_stream_get_formats(resolved_stream))) {
+				ast_stream_free(resolved_stream); /* Can handle NULL */
+				if (session->endpoint->media.codec_prefs_incoming_answer.transcode == CODEC_NEGOTIATION_TRANSCODE_ALLOW) {
+					ast_log(LOG_DEBUG, "%s: No common codecs between channels and transcoding allowed.\n",
+						ast_sip_session_get_name(session));
+					skip_stream_setting = 1;
+				} else {
+					ast_log(LOG_WARNING, "%s: No common codecs between channels and transcoding not allowed.\n",
+						ast_sip_session_get_name(session));
+					return -1;
+				}
+			}
+			if (!skip_stream_setting && !ast_format_cap_identical(ast_stream_get_formats(resolved_stream), ast_stream_get_formats(configured_stream))) {
+				ast_log(LOG_TRACE, "%s: Replacing pending stream: %s with resolved stream: %s\n", ast_sip_session_get_name(session),
+					ast_str_tmp(256, ast_stream_to_str(configured_stream, &STR_TMP)),
+					ast_str_tmp(256, ast_stream_to_str(resolved_stream, &STR_TMP)));
+				if (ast_stream_topology_set_stream(session->pending_media_state->topology, index, resolved_stream)) {
+					ast_stream_free(resolved_stream);
+					ast_log(LOG_WARNING, "%s: Couldn't set stream in topology.\n",
+						ast_sip_session_get_name(session));
+					return -1;
+				}
+			}
```

Comment:
resolved_stream is never freed when skip_stream_setting is true or when resolved_stream is identical to configured_stream, there should be an else branch with ``ast_stream_free(resolved_stream);`` here.

Reply:
Thank you! I have fixed the leak and refactored the logic to make it simpler. Please mark as resolved if it looks good to you.